// vuepress-theme-hope@2.0.0-rc.106/dist/client/components/navbar/Navbar.js
import { hasGlobalComponent } from "@vuepress/helper/client";
import noopComponent from "@vuepress/helper/noopComponent";
import { computed, defineComponent, h, ref, resolveComponent, watch } from "vue";
import { onContentUpdated } from "vuepress/client";
import AppearanceButton from "@theme-hope/components/appearance/AppearanceButton";
import LanguageDropdown from "@theme-hope/components/navbar/LanguageDropdown";
import NavbarBrand from "@theme-hope/components/navbar/NavbarBrand";
import NavbarLinks from "@theme-hope/components/navbar/NavbarLinks";
import NavScreen from "@theme-hope/components/navbar/NavScreen";
import RepoLink from "@theme-hope/components/navbar/RepoLink";
import ToggleNavbarButton from "@theme-hope/components/navbar/ToggleNavbarButton";
import ToggleSidebarButton from "@theme-hope/components/navbar/ToggleSidebarButton";
import { useNavbarAutoHide } from "@theme-hope/composables/navbar/useNavbarAutoHide";
import { useThemeLocale } from "@theme-hope/composables/useTheme";
import { useWindowSize } from "@theme-hope/composables/useWindowSize";
import "../../styles/navbar/navbar.scss";
export default defineComponent({
    name: "NavBar",
    emits: ["toggleSidebar"],
    slots: Object,
    setup(_props, { emit, slots }) {
        const themeLocale = useThemeLocale();
        const { isMobile } = useWindowSize();
        const showScreen = ref(false);
        const autoHide = useNavbarAutoHide();
        const navbarLayout = computed(() => themeLocale.value.navbarLayout ??
            {
                start: ["Brand"],
                center: ["Links"],
                end: ["Language", "Repo", "Outlook", "Search"],
            });
        const navbarComponentMap = {
            Brand: NavbarBrand,
            Language: __VP_I18N__ ? LanguageDropdown : noopComponent,
            Links: NavbarLinks,
            Repo: RepoLink,
            Outlook: AppearanceButton,
            Search: hasGlobalComponent("SearchBox") ? resolveComponent("SearchBox") : noopComponent,
        };
        const getNavbarComponent = (component) => navbarComponentMap[component] ??
            (hasGlobalComponent(component) ? resolveComponent(component) : noopComponent);
        onContentUpdated(() => {
            showScreen.value = false;
        });
        watch(isMobile, (value) => {
            if (!value)
                showScreen.value = false;
        });
        return () => [
            h("header", {
                key: "navbar",
                id: "navbar",
                class: ["vp-navbar", { "auto-hide": autoHide.value }],
                "vp-navbar": "",
            }, [
                h("div", { class: "vp-navbar-start" }, [
                    h(ToggleSidebarButton, {
                        onToggle: () => {
                            if (showScreen.value)
                                showScreen.value = false;
                            emit("toggleSidebar");
                        },
                    }),
                    navbarLayout.value.start?.map((item) => h(getNavbarComponent(item))),
                ]),
                h("div", { class: "vp-navbar-center" }, [
                    navbarLayout.value.center?.map((item) => h(getNavbarComponent(item))),
                ]),
                h("div", { class: "vp-navbar-end" }, [
                    navbarLayout.value.end?.map((item) => h(getNavbarComponent(item))),
                    h(ToggleNavbarButton, {
                        active: showScreen.value,
                        onToggle: () => {
                            showScreen.value = !showScreen.value;
                        },
                    }),
                ]),
            ]),
            h(NavScreen, { show: showScreen.value }, slots),
        ];
    },
});
