// vuepress-theme-hope@2.0.0-rc.106/dist/client/utils/info/resolveEditLink.js
import { isLinkHttp, removeEndingSlash, removeLeadingSlash } from "@vuepress/helper/client";
import { resolveRepoType } from "vuepress-shared/client";
export const editLinkPatterns = {
    GitHub: ":repo/edit/:branch/:path",
    GitLab: ":repo/-/edit/:branch/:path",
    Gitee: ":repo/edit/:branch/:path",
    Bitbucket: ":repo/src/:branch/:path?mode=edit&spa=0&at=:branch&fileviewer=file-view-default",
};
export const resolveEditLink = ({ docsRepo, docsBranch, docsDir, filePathRelative, editLinkPattern, }) => {
    if (!filePathRelative)
        return null;
    const repoType = resolveRepoType(docsRepo);
    let pattern = "";
    if (editLinkPattern)
        pattern = editLinkPattern;
    else if (repoType != null)
        pattern = editLinkPatterns[repoType];
    if (!pattern)
        return null;
    return pattern
        .replace(/:repo/u, isLinkHttp(docsRepo) ? docsRepo : `https://github.com/${docsRepo}`)
        .replace(/:branch/u, docsBranch)
        .replace(/:path/u, removeLeadingSlash(`${removeEndingSlash(docsDir)}/${filePathRelative}`));
};
